### PR TITLE
Use the site search endpoint for plugin search

### LINKED ISSUE #62729: Marketplace: Update WP.org plugin search to use ES endpoint from Calypso
As part of pdh6GB-Ax-p2 we can use `https://public-api.wordpress.com/rest/v1/sites/108986944/search` by adding any required post parameters to get the .org plugins.

Fields needed:

- [x] `active_installs`
- [x] `author`
- [ ] `author_profile`
- [ ] `banners` or `image`
- [ ] `contributors`
- [ ] `icon`
- [ ] `last_updated`
- [ ] `name`
- [ ] `rating`
- [ ] `description`
- [ ] `short_description`
- [ ] `faq`
- [ ] `installation`
- [ ] `slug`
- [ ] `tags`
- [ ] `version`
- [ ] `tested`

### PR DESCRIPTION
**Pull request by gcsecsey**
### Changes proposed in this Pull Request

* Adds a new hook to use the `https://public-api.wordpress.com/rest/v1/sites/108986944/search` endpoint
* Adds the same parameters for author and category search as WPORG
* Moves some common code to utils and constants files

### Testing instructions

* Visit `/plugins`
* Search for something generic (eg. `wordpress`)
* Verify that both WPCOM and WPORG results are loaded
* Verify that WPORG results have all the necessary information on the search result card (title, excerpt, rating, last update, number of installs)
* Scroll to the bottom, and verify that the search result list loads the next page of results 

**Videos**
Plugin search: https://d.pr/v/bvi2uz
Author search: https://d.pr/v/9ntpmF

#### Category search

The index contains only WPORG categories, and we can search for those. However, the categories used on the WPORG API (Popular, Featured), are stored in the `plugin_section` taxonomy and are not in the index response: 
https://github.com/WordPress/wordpress.org/blob/38e38a5b0bbca67affaf0ec70617db14f3f88f84/wordpress.org/public_html/wp-content/plugins/plugin-directory/class-plugin-directory.php#L195
https://github.com/WordPress/wordpress.org/blob/38e38a5b0bbca67affaf0ec70617db14f3f88f84/wordpress.org/public_html/wp-content/plugins/plugin-directory/class-plugin-directory.php#L760

We should assess if we need to add this taxonomy to the index as well.

#### Localization

The current implementation only supports results in english. The index stores all the translations in the MAG16, so we can create the logic of picking the correct keys in a follow-up PR.

#### Icons

The index contains icon file names in a serialized PHP array. These are deserialized and appended to the WPORG image host (`https://ps.w.org`). For eg. Woocommerce plugins, the PHP array is empty, so icon URLs can't be produced just from the index response. We'll need to find an alternative, more stable way of getting the icon URLs.

**To reproduce:** try searching for `WooTcsShipmentMaker`

| Search results (index) | Plugin details page (WPORG) |
| ------ | ----- |
| image | image |

#### Author profile link

The index response doesn't contain the `author_profile` URL, which is a WPORG profile URL. I'm not sure we need it though, as the author name currently redirects to the author search page, and not the `author_profile` URL. It's also only used on the plugin details page, not the search results page.

| Search results | Plugin details page |
| ------ | ----- |
| image | image |

#### Plugin details page

We can query for an exact match on the product slug from the index, which would let us get the same data to display on the plugin details page. However, as the plugin details page uses more fields from the WPORG response, we should investigate in a follow-up issue whether we need all the required fields to display the plugin page.

Related to #62729

### CODE REVIEW

**Comment by lsl** [comment]
Will this somehow make use of this new flag? 12392-gh-Automattic/woocommerce.com#issuecomment-1105140159

**Comment by gcsecsey** [comment]
> Will this somehow make use of this new flag? [12392-gh-Automattic/woocommerce.com#issuecomment-1105140159](https://href.li/?12392-gh-Automattic/woocommerce.com#issuecomment-1105140159)

@lsl the MVP version only supports WPORG results, so I don't think we'll be able to filter for `marketplace_categories` just yet. D79130-code is adding WPCOM and WCCOM results to a new index and would let us filter based on taxonomies.

**Comment by WBerredo** [comment]
**Author profile link** -> currently this we define the profile keyword this way: https://github.com/Automattic/wp-calypso/blob/6406aceea9155168d38ab47ba7c18acbaa346678/client/lib/plugins/utils.js#L284-L288

File: client/data/marketplace/constants.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export const BASE_STALE_TIME = 1000 * 60 * 60 * 2; // 2 hours
+export const DEFAULT_PAGE_SIZE = 24;
+export const DEFAULT_FIRST_PAGE = 1;
+export const DEFAULT_CATEGORY = 'all';
+
+export const SITE_SEARCH_ENDPOINT =
+	'https://public-api.wordpress.com/rest/v1/sites/108986944/search';
```
Comment (WBerredo) [inline client/data/marketplace/constants.ts]: Any chance you add a comment or a different constant explaining this value `108986944`?

File: client/data/marketplace/use-site-search-es-query.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+		revision: string;
+	};
+	const icons: Record< string, Icon > = phpUnserialize( iconsArray );
+	return Object.values( icons ).reduce(
+		( prev: Record< string, string >, { resolution, filename, location, revision } ) => {
+			prev[
+				resolution
+			] = `https://ps.w.org/${ pluginSlug }/${ location }/${ filename }?rev=${ revision }`;
+			return prev;
+		},
```
Comment (WBerredo) [inline client/data/marketplace/use-site-search-es-query.ts:40]: Will we always have those 4 parameters? (resolution, filename, location, revision)

Do you know if there is any kind of fallback for the times we don't have some of them?

File: client/data/marketplace/use-site-search-es-query.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+		{}
+	);
+};
+
+const createAuthorUrl = ( header_author: string, header_author_uri: string ) =>
+	`<a href="${ header_author_uri }">${ header_author }</a>`;
```
Comment (WBerredo) [inline client/data/marketplace/use-site-search-es-query.ts]: Nitpicking: I would use camelCase when we are not explicitly using a fiedName.

```suggestion
const createAuthorUrl = ( headerAuthor: string, headerAuthorUri: string ) =>
	`<a href="${ headerAuthorUri }">${ headerAuthor }</a>`;
```

File: client/data/marketplace/use-site-search-es-query.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+const mapIndexResultsToPluginData = ( results: ESHits ): Plugin[] => {
+	return results.map( ( { fields: hit } ) => {
+		return {
+			name: hit.title_en, // TODO: add localization
+			slug: hit.slug,
+			version: hit.stable_tag,
+			author: createAuthorUrl(
+				hit[ 'meta.header_author.value' ],
+				hit[ 'meta.header_author_uri.value' ]
+			),
```
Comment (WBerredo) [inline client/data/marketplace/use-site-search-es-query.ts:59]: Do we have a `contributors` field? 

We saw we were getting better results when using it to find the author for search.

Reference: https://github.com/Automattic/wp-calypso/pull/62777#issuecomment-1109980813

File: client/data/marketplace/use-site-search-es-query.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+		headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
+	} );
+
+	if ( response.ok ) {
+		return await response.json();
+	}
+	throw new Error( await response.text() );
+}
+
+export async function fetchWPOrgPluginsFromSiteSearch( options: PluginQueryOptions ) {
```
Comment (WBerredo) [inline client/data/marketplace/use-site-search-es-query.ts]: Nitpicking

The ES index will have only WPOrg plugins for now?

Maybe it is a good idea to use a more generalized method name as it should come from the same place(unified index) in the future, and terms as WPcom and WPorg plugins won't make sense anymore.

File: client/data/marketplace/use-site-search-es-query.ts [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+		size,
+		from,
+	} );
+
+	return requestResult.results;
+}
+
+const extractPages = ( pages: Array< { hits: ESHits } > = [] ) => {
+	const allHits = pages.flatMap( ( page ) => page.hits );
+	return normalizePluginsList( mapIndexResultsToPluginData( allHits ) );
```
Comment (WBerredo) [inline client/data/marketplace/use-site-search-es-query.ts:255]: Nitpicking: I think we can simplify this piece in the future. We are mapping the ES results to look like the old requests and then we normalize this data for usage in the application. I would say we could do this straight from ES results to application-ready data.

That said, I think it should be a follow-up PR as we are changing a lot already at this one.

**Review by WBerredo — DISMISSED** [review]
Before merging this PR, we must discuss if we will add an experiment to compare the results of both experiences. 

In a positive case, we would need some work setting up this experiment and some small changes to `PluginsBrowser` component.

**Review by cpapazoglou — DISMISSED** [review]
Thanks for working on this @gcsecsey! I works quite good and the API response is 25% faster 👏👏 👏 

Some feedback:
1. Let's use a feature flag and conditionally (manually) enable it in all environments. This will allow us to push the infrastructure changes and then iterate on the details.
2. Ratings need some different parsing as it seems
CleanShot 2022-04-27 at 11 05 03@2x
3. We will definitely need handle author search better as @WBerredo suggested
4. It seems that sorting is a tad different 

|Before | After|
|-------|------|
|||

I guess this is `active_installations` related. Can we sort by this property?

File: client/data/marketplace/use-site-search-es-query.ts (hunk 2 above)
Comment (gcsecsey) [reply client/data/marketplace/use-site-search-es-query.ts:40]: The icons array contains these keys after converted to JSON: 

```json
{
  "icon-128x128.png": 
    "filename": "icon-128x128.png"
    "locale": ""
    "location": "assets"
    "resolution": "128x128"
    "revision": 2366418
}
```

So far, all results contained those 4 keys, and I couldn't find any results where the `locale` key is used. As this is originally stored as a PHP array, I wasn't able to filter for it in ES.

I don't know what could be a sensible fallback for these, as the index doesn't contain anything else regarding icons. For now, the `reduce` returns an empty array if there are no icons in the `assets` folder, and the default icon is used on the UI.

File: client/data/marketplace/constants.ts (hunk 1 above)
Comment (gcsecsey) [reply client/data/marketplace/constants.ts]: Good idea to make it more clear, I've added it as a constant! 🙂

File: client/data/marketplace/use-site-search-es-query.ts (hunk 5 above)
Comment (gcsecsey) [reply client/data/marketplace/use-site-search-es-query.ts]: AFAIK yes, until D79130-code is landed we won't be able to query other sources.

I agreee that the name should be more generic, I've updated it. 👍

File: client/data/marketplace/use-site-search-es-query.ts (hunk 6 above)
Comment (gcsecsey) [reply client/data/marketplace/use-site-search-es-query.ts:255]: I agree that the mapping could be further simplified, but I wanted to touch as few layers as possible with the MVP. My original idea was to replace only the fetch layer within the react-query and do a data-mapping that would let the `select` pick up the response the same way. Unfortunately, the responses are so different (e.g pagination for the ES response has to be calculated from the query params and the `total` number in the response), that I had to create a new query after all.

Later, when all results can be queried from the same index, we can definitely come up with a simpler mapping and remove a lot of redundant logic.

File: client/data/marketplace/use-site-search-es-query.ts (hunk 4 above)
Comment (gcsecsey) [reply client/data/marketplace/use-site-search-es-query.ts:59]: I've tried using the function, or implementing it again for the ES index. The index has something similar, the problem is that the data is in a different structure:

Each hit contains a `taxonomy.plugin_contributors` and a flat `contributors` array:
```json
{
    "taxonomy": {
        "plugin_contributors": [
            {
                "name": "automattic",
                "slug": "automattic",
                "term_id": "77494"
            },
            {
                "name": "mikejolley",
                "slug": "mikejolley",
                "term_id": "77602"
            }
        ]
    },
    "contributors": [
        "automattic",
        "mikejolley"
    ]
}
```

Also, the `meta.header_author` object contains a display name, while the rest of the fields contain the slug/username. In the case of Automattic it's just an upper/lower case difference, but there are authors for which these two are entirely different. Without a slug field to match on, it's hard to choose the correct contributor as the author from the list.
```json
"header_author": {
    "value": [
        "Automattic"
    ],
    "long": [
        0
    ],
    "double": [
        0
    ],
    "boolean": [
        true
    ]
},
```

I'll look into how the contributors array is ordered, maybe the header author is always the first contributor.

File: client/data/marketplace/use-site-search-es-query.ts [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+						field: 'rating',
+						factor: 0.25,
+						modifier: 'sqrt',
+						missing: 2.5,
+					},
+				},
+			],
+			score_mode: 'multiply',
+		},
+	};
```
Comment (gcsecsey) [inline client/data/marketplace/use-site-search-es-query.ts:168]: I've tried to port the sorting criteria that is sent to the site search endpoint from the WPORG backend here: https://github.com/WordPress/wordpress.org/blob/7f3e0d50e02008f5edd46d281052caf827661550/wordpress.org/public_html/wp-content/plugins/plugin-directory/libs/site-search/jetpack-search.php#L1040

It still doesn't match the original results, but results with a high number of installations and recent updates are given higher scores.

| Before | After | WPORG API |
| ------ | ----- | --- | 
| image | image | image |

File: client/data/marketplace/use-site-search-es-query.ts (hunk 2 above)
Comment (WBerredo) [reply client/data/marketplace/use-site-search-es-query.ts:40]: At first, I was thinking about two possible fallbacks:
* resolution: but we should be handling it later down the line when we choose which version to pick
* revision: I was wondering it was sequential like `1, 2, 3, etc`, so we could use a fallback to `1`. Thanks for adding a response example here as I was not guessing it right.

That said, I think the current version is fine. Ty.

Comment (gcsecsey) [reply client/data/marketplace/use-site-search-es-query.ts:40]: For the plugins that have icon assets uploaded to WPORG, these fields are always indexed. We have to have a fallback for plugins that don't have icons uploaded, I've just found out that WPORG generates a pattern for these (eg. [here](https://wordpress.org/plugins/slideshare/)). I'll incorporate a similar fallback in a follow-up PR.

**Comment by gcsecsey** [comment]
I've added the `marketplace-jetpack-plugin-search` feature flag, it is disabled by default on all environments.

To use the site index search endpoint instead of the WPORG API when searching, the feature flag can be enabled by:
- running `ENABLE_FEATURES=marketplace-jetpack-plugin-search yarn start`
- adding `?flags=marketplace-jetpack-plugin-search` at the end of the URL

**Comment by gcsecsey** [comment]
I'll move all blockers identified so far to [the issue of this PR](https://github.com/Automattic/wp-calypso/issues/62729). As this feature is now behind a feature flag, we can merge and deploy this PR.

**Review by cpapazoglou — APPROVED** [review]
- ✅ Uses the default method (.org API) when not feature flag is added.
- ✅ Uses the elastic search when `?flags=marketplace-jetpack-plugin-search` is added
- ✅ Feature is by default false in all environments.

Good work @gcsecsey 👏 ! Let's 🚢  this beast and iterate on it.

**Pull request merged**